chain-impl-mockchain/src/update.rs

```diff
@@ -1,94 +1,75 @@
 //use crate::certificate::{verify_certificate, HasPublicKeys, SignatureRaw};
+use crate::certificate::{UpdateProposal, UpdateProposalId, UpdateVote, UpdateVoterId};
 use crate::date::BlockDate;
-use crate::fragment::config::ConfigParams;
-use crate::key::{deserialize_signature, serialize_signature, verify_signature, BftLeaderId};
 use crate::setting::{ActiveSlotsCoeffError, Settings};
-use chain_core::mempack::{ReadBuf, ReadError, Readable};
-use chain_core::property;
-use chain_crypto::{Ed25519, Signature, Verification};
-use std::collections::{HashMap, HashSet};
+use imhamt::Hamt;
+use std::collections::hash_map::DefaultHasher;
 
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct UpdateState {
-    pub proposals: HashMap<UpdateProposalId, UpdateProposalState>,
+    pub proposals: Hamt<DefaultHasher, UpdateProposalId, UpdateProposalState>,
```

Comment:
Just a question. Why do we move from the std::HashMap to the Hamt as I see it is a custom implementation of the Hash map ? What is the difference between them ?

Reply:
haven't read the rest of the PR, but the hamt is immutable (or persistent), which means it can be cloned cheaply. A big chunk of the ledger uses it so it can be cloned once per block without doing a deep clone.